src/v/cluster/rm_stm.cc

```diff
@@ -1986,6 +1990,38 @@ void rm_stm::fill_snapshot_wo_seqs(T& snapshot) {
 }
 
 ss::future<stm_snapshot> rm_stm::take_snapshot() {
+    auto start_offset = _raft->start_offset();
+
+    std::vector<abort_index> abort_indexes;
+    std::vector<abort_index> expired_abort_indexes;
+    abort_indexes.reserve(_log_state.abort_indexes.size());
+
+    for (const auto& idx : _log_state.abort_indexes) {
+        if (idx.last < start_offset) {
+            // caching expired indexes instead of removing them as we go
+            // to avoid giving a control to another coroutine and managing
+            // concurrent access to _log_state.abort_indexes
+            expired_abort_indexes.push_back(idx);
+        } else {
+            abort_indexes.push_back(idx);
+        }
+    }
+    _log_state.abort_indexes = std::move(abort_indexes);
```

Comment:
I see `_log_state.abort_indexes` access in many places, such the iteration here:

```c++
ss::future<> rm_stm::remove_persistent_state() {
    for (const auto& idx : _log_state.abort_indexes) {
        auto filename = abort_idx_name(idx.first, idx.last);
        co_await _abort_snapshot_mgr.remove_snapshot(filename);
    }
```

it looks like if `_log_state.abort_indexes = std::move(abort_indexes);` were interleaved with the iteration loop (the loop has a co_await scheduling point) that this would result in a seg fault.

what prevents this interleaving and interleavings with other places where `_log_state.{aborted, abort_indexes}` are accessed?

Reply:
awesome catch, I've added a couple commits to enforce atomic access

Reply:
nice can those commits be squashed into here?